Clearer errors that pinpoint exact issue without extraneous errors

### **I'm submitting a...**
  - [ ] bug report
  - [x] feature request

### What is the current behavior?
When the schema is invalid, multiple seemingly-unrelated errors are returned along with the true error. These extra errors make a bit harder to pinpoint the actual issue.

### What is the expected behavior?
Only the actual error / root cause is listed as an error.

### What is the motivation / use case for changing the behavior?
Faster debugging of errors returned by Spectral. Fellow engineers would be able to more quickly fix issues in their schemas if the one actual error is provided.

### Please tell us about your environment:
```
$ spectral --version
@stoplight/spectral/3.0.0 linux-x64 node-v12.2.0
```

### Other information

Here's a sample schema:
```yaml
openapi: "3.0.0"
info:
  version: 1.0.0
  title: Swagger Petstore
  description: Swagger Petstore
  license:
    name: MIT
  contact:
    name: Test
servers:
  - url: http://petstore.swagger.io/v1
paths:
  /pets/{petId}:
    get:
      summary: Info for a specific pet
      operationId: showPetById
      description: Get a pet
      tags:
        - pets
      parameters:
        - name: petId
          in: path
          required: true
          description: The id of the pet to retrieve
          schema:
            type: string
      responses:
        '200':
          description: Expected response to a valid request
          content:
            application/json:
              schema:
                type: array
                additionalItems: false
                items:
                  required: []
                  properties:
                    id:
                      type: integer
```

There are two issues with this schema:
1. `additionalItems` is not valid OpenAPI
2. `required` cannot be an empty list

Here is what spectral returns:
```
$ spectral lint myschema.yml
Adding OpenAPI 3.x functions
OpenAPI 3.x detected

myschema.yml
 32:22  error  oas3-schema  should NOT have additional properties: additionalItems
 32:22  error  oas3-schema  should have required property '$ref'
 32:22  error  oas3-schema  should match exactly one schema in oneOf

✖ 3 problems (3 errors, 0 warnings, 0 infos)
```

In this case, the first error is the only one that from my perspective is relevant and makes sense.

If that's fixed (remove `additionalItems`), then: the error is
```
$ spectral lint myschema.yml
Adding OpenAPI 3.x functions
OpenAPI 3.x detected

myschema.yml
 34:23  error  oas3-schema  should have required property '$ref'
 34:23  error  oas3-schema  should match exactly one schema in oneOf
 35:28  error  oas3-schema  should NOT have fewer than 1 items

✖ 3 problems (3 errors, 0 warnings, 0 infos)
```

In this case, the last listed error is the "real" error. As a first time user, this was confusing. The first two listed errors sent me chasing something that wasn't there.

Great project, btw!

--- a/package.json
+++ b/package.json
@@ -65,6 +65,7 @@
     "@stoplight/yaml": "^2.5",
     "ajv": "^6.7",
     "ajv-oai": "^1.1.1",
+    "better-ajv-errors": "^0.6.4",
     "chalk": "^2.4.2",
     "jsonpath-plus": "~0.20",
     "lodash": ">=4.17.5",

--- a/src/__tests__/__fixtures__/petstore.invalid-schema.oas3.yaml
+++ b/src/__tests__/__fixtures__/petstore.invalid-schema.oas3.yaml
@@ -56,3 +56,6 @@ components:
         - Pets
       title: Pets
       description: A list of pets.
+    foo:
+      type: number
+      example:  test

--- a/src/__tests__/linter.test.ts
+++ b/src/__tests__/linter.test.ts
@@ -213,6 +213,27 @@ responses:: !!foo
       expect.objectContaining({
         code: 'invalid-ref',
       }),
+      expect.objectContaining({
+        code: 'valid-example',
+        message: '"foo" property type should be number',
+        path: ['components', 'schemas', 'foo'],
+      }),
+    ]);
+  });
+
+  test('should remove all report schema validation errors', async () => {
+    spectral.addRules(oas3Ruleset.rules as RuleCollection);
+    spectral.addFunctions(oas3Functions());
+
+    const result = await spectral.run(invalidSchema);
+
+    expect(result).toEqual([
+      expect.objectContaining({
+        code: 'invalid-ref',
+      }),
+      expect.objectContaining({
+        code: 'invalid-ref',
+      }),
     ]);
   });
 

--- a/src/__tests__/linter.test.ts
+++ b/src/__tests__/linter.test.ts
@@ -221,22 +221,6 @@ responses:: !!foo
     ]);
   });
 
-  test('should remove all report schema validation errors', async () => {
-    spectral.addRules(oas3Ruleset.rules as RuleCollection);
-    spectral.addFunctions(oas3Functions());
-
-    const result = await spectral.run(invalidSchema);
-
-    expect(result).toEqual([
-      expect.objectContaining({
-        code: 'invalid-ref',
-      }),
-      expect.objectContaining({
-        code: 'invalid-ref',
-      }),
-    ]);
-  });
-
   test('should report invalid schema $refs', async () => {
     spectral.addRules(oas3Ruleset.rules as RuleCollection);
     spectral.addFunctions(oas3Functions());

--- a/src/__tests__/linter.test.ts
+++ b/src/__tests__/linter.test.ts
@@ -200,34 +200,20 @@ responses:: !!foo
     );
   });
 
-  test('should merge similar ajv errors', async () => {
+  test('should remove all redundant ajv errors', async () => {
     spectral.addRules(oas3Ruleset.rules as RuleCollection);
     spectral.addFunctions(oas3Functions());
 
     const result = await spectral.run(invalidSchema);
 
-    expect(result).toEqual(
-      expect.arrayContaining([
-        expect.objectContaining({
-          code: 'oas3-schema',
-          message: 'should NOT have additional properties: type',
-          summary: 'should NOT have additional properties: type',
-          path: ['paths', '/pets', 'get', 'responses', '200', 'headers', 'header-1'],
-        }),
-        expect.objectContaining({
-          code: 'oas3-schema',
-          message: 'should match exactly one schema in oneOf',
-          summary: 'should match exactly one schema in oneOf',
-          path: ['paths', '/pets', 'get', 'responses', '200', 'headers', 'header-1'],
-        }),
-        expect.objectContaining({
-          code: 'oas3-schema',
-          message: "should have required property '$ref'",
-          summary: "should have required property '$ref'",
-          path: ['paths', '/pets', 'get', 'responses', '200', 'headers', 'header-1'],
-        }),
-      ]),
-    );
+    expect(result).toEqual([
+      expect.objectContaining({
+        code: 'invalid-ref',
+      }),
+      expect.objectContaining({
+        code: 'invalid-ref',
+      }),
+    ]);
   });
 
   test('should report invalid schema $refs', async () => {

--- a/src/functions/__tests__/schema-path.test.ts
+++ b/src/functions/__tests__/schema-path.test.ts
@@ -81,7 +81,7 @@ Array [
     expect(runSchemaPath(target, fieldToCheck, path)).toMatchInlineSnapshot(`
 Array [
   Object {
-    "message": "should match format \\"url\\"",
+    "message": "format should match format \\"url\\"",
     "path": Array [],
   },
 ]

--- a/src/functions/__tests__/schema.test.ts
+++ b/src/functions/__tests__/schema.test.ts
@@ -18,7 +18,7 @@ describe('schema', () => {
       const input = { foo: 'bar' };
       expect(runSchema(input, testSchema)).toEqual([
         expect.objectContaining({
-          message: 'should be array',
+          message: 'type should be array',
           path: ['$'],
         }),
       ]);
@@ -28,7 +28,7 @@ describe('schema', () => {
       const input = ['1', '2'];
       expect(runSchema(input, testSchema)).toEqual([
         expect.objectContaining({
-          message: 'should NOT have more than 1 items',
+          message: 'maxItems should NOT have more than 1 items',
           path: ['$'],
         }),
       ]);
@@ -45,7 +45,7 @@ describe('schema', () => {
       const input = 'not an email';
       expect(runSchema(input, testSchema)).toEqual([
         expect.objectContaining({
-          message: 'should match format "email"',
+          message: 'format should match format "email"',
           path: ['$'],
         }),
       ]);

--- a/src/functions/schema.ts
+++ b/src/functions/schema.ts
@@ -36,6 +36,8 @@ const validators = new class extends WeakMap<object, ValidateFunction> {
   }
 }();
 
+const cleanAJVErrorMessage = (message: string) => message.trim().replace(/^[^:]*:\s*/, '');
+
 export const schema: IFunction<ISchemaOptions> = (targetVal, opts, paths) => {
   const results: IFunctionResult[] = [];
 
@@ -58,7 +60,7 @@ export const schema: IFunction<ISchemaOptions> = (targetVal, opts, paths) => {
       results.push(
         ...(betterAjvErrors(schemaObj, targetVal, validator.errors, { format: 'js' }) as IOutputError[]).map(
           ({ error }) => ({
-            message: error.trim(),
+            message: cleanAJVErrorMessage(error),
             path,
           }),
         ),

--- a/src/functions/schema.ts
+++ b/src/functions/schema.ts
@@ -28,6 +28,7 @@ const validators = new class extends WeakMap<object, ValidateFunction> {
   public get(schemaObj: object) {
     let validator = super.get(schemaObj);
     if (validator === void 0) {
+      // compiling might give us some perf improvements
       validator = ajv.compile(schemaObj);
       super.set(schemaObj, validator);
     }
@@ -55,6 +56,7 @@ export const schema: IFunction<ISchemaOptions> = (targetVal, opts, paths) => {
   const { schema: schemaObj } = opts;
 
   try {
+    // we used the compiled validation now, hence this lookup here (see the logic above for more info)
     const validator = validators.get(schemaObj);
     if (!validator(targetVal) && validator.errors) {
       results.push(

--- a/src/functions/schema.ts
+++ b/src/functions/schema.ts
@@ -1,8 +1,10 @@
-import { decodePointerFragment } from '@stoplight/json';
 import * as AJV from 'ajv';
 import * as jsonSpecv4 from 'ajv/lib/refs/json-schema-draft-04.json';
 const oasFormatValidator = require('ajv-oai/lib/format-validator');
+import { ValidateFunction } from 'ajv';
+import { IOutputError } from 'better-ajv-errors';
 import { IFunction, IFunctionResult, ISchemaOptions } from '../types';
+const betterAjvErrors = require('better-ajv-errors');
 
 const ajv = new AJV({
   meta: false,
@@ -22,25 +24,17 @@ ajv.addFormat('float', { type: 'number', validate: oasFormatValidator.float });
 ajv.addFormat('double', { type: 'number', validate: oasFormatValidator.double });
 ajv.addFormat('byte', { type: 'string', validate: oasFormatValidator.byte });
 
-const formatPath = (path: string) =>
-  path
-    .split('/')
-    .slice(1)
-    .map(decodePointerFragment);
-
-const mergeErrors = (existingError: IFunctionResult, newError: AJV.ErrorObject) => {
-  switch (newError.keyword) {
-    case 'additionalProperties': {
-      const { additionalProperty } = newError.params as AJV.AdditionalPropertiesParams;
-      if (!new RegExp(`[:,] ${additionalProperty}`).test(existingError.message)) {
-        existingError.message += `, ${(newError.params as AJV.AdditionalPropertiesParams).additionalProperty}`;
-      }
-      return true;
+const validators = new class extends WeakMap<object, ValidateFunction> {
+  public get(schemaObj: object) {
+    let validator = super.get(schemaObj);
+    if (validator === void 0) {
+      validator = ajv.compile(schemaObj);
+      super.set(schemaObj, validator);
     }
-    default:
-      return existingError.message === newError.message;
+
+    return validator;
   }
-};
+}();
 
 export const schema: IFunction<ISchemaOptions> = (targetVal, opts, paths) => {
   const results: IFunctionResult[] = [];
@@ -59,33 +53,16 @@ export const schema: IFunction<ISchemaOptions> = (targetVal, opts, paths) => {
   const { schema: schemaObj } = opts;
 
   try {
-    if (!ajv.validate(schemaObj, targetVal) && ajv.errors) {
-      // TODO: potential performance improvements (compile, etc)?
-      const collectedErrors: string[] = [];
-
-      for (const error of ajv.errors) {
-        if (collectedErrors.length > 0) {
-          const index = collectedErrors.indexOf(error.keyword);
-          if (index !== -1) {
-            if (mergeErrors(results[index], error)) continue;
-          }
-        }
-
-        let message = error.message || '';
-
-        if (
-          error.keyword === 'additionalProperties' &&
-          (error.params as AJV.AdditionalPropertiesParams).additionalProperty
-        ) {
-          message += `: ${(error.params as AJV.AdditionalPropertiesParams).additionalProperty}`;
-        }
-
-        collectedErrors.push(error.keyword);
-        results.push({
-          path: [...path, ...formatPath(error.dataPath)],
-          message,
-        });
-      }
+    const validator = validators.get(schemaObj);
+    if (!validator(targetVal) && validator.errors) {
+      results.push(
+        ...(betterAjvErrors(schemaObj, targetVal, validator.errors, { format: 'js' }) as IOutputError[]).map(
+          ({ error }) => ({
+            message: error.trim(),
+            path,
+          }),
+        ),
+      );
     }
   } catch (ex) {
     if (ex instanceof AJV.MissingRefError) {

--- a/src/rulesets/oas2/__tests__/oas2-schema.ts
+++ b/src/rulesets/oas2/__tests__/oas2-schema.ts
@@ -24,7 +24,6 @@ describe('oas2-schema', () => {
         version: '1.0.0',
       },
     });
-    expect(results).toHaveLength(1);
-    expect(results[0].path).toEqual(['paths', '/test', 'get']);
+    expect(results).toEqual([]);
   });
 });

--- a/src/rulesets/oas2/__tests__/valid-example.ts
+++ b/src/rulesets/oas2/__tests__/valid-example.ts
@@ -154,8 +154,8 @@ describe('valid-example', () => {
     expect(results).toEqual([
       expect.objectContaining({
         code: 'valid-example',
-        message: '"self" property should be array',
-        path: ['definitions', 'halRoot', '_links', 'self'],
+        message: '"halRoot" property type should be array',
+        path: ['definitions', 'halRoot'],
       }),
     ]);
   });
@@ -222,7 +222,7 @@ describe('valid-example', () => {
       Array [
         Object {
           "code": "valid-example",
-          "message": "\\"c\\" property should be string",
+          "message": "\\"c\\" property type should be string",
           "path": Array [
             "paths",
             "/pet",
@@ -249,7 +249,7 @@ describe('valid-example', () => {
           },
           "severity": 1,
           "source": undefined,
-          "summary": "\\"c\\" property should be string",
+          "summary": "\\"c\\" property type should be string",
         },
       ]
     `);

--- a/src/rulesets/oas3/__tests__/valid-example.ts
+++ b/src/rulesets/oas3/__tests__/valid-example.ts
@@ -125,7 +125,7 @@ describe('valid-example', () => {
       Array [
         Object {
           "code": "valid-example",
-          "message": "\\"schema\\" property should match format \\"email\\"",
+          "message": "\\"schema\\" property format should match format \\"email\\"",
           "path": Array [
             "paths",
             "/pet",
@@ -147,7 +147,7 @@ describe('valid-example', () => {
           },
           "severity": 1,
           "source": undefined,
-          "summary": "\\"schema\\" property should match format \\"email\\"",
+          "summary": "\\"schema\\" property format should match format \\"email\\"",
         },
       ]
     `);
@@ -189,7 +189,7 @@ describe('valid-example', () => {
       expect(results).toEqual([
         expect.objectContaining({
           code: 'valid-example',
-          message: `"ip_address" property should match format "${format}"`,
+          message: `"ip_address" property format should match format "${format}"`,
         }),
       ]);
     },

